### ISSUE TITLE
Manuscript Build Settings - Close not asking for confirmation after changes (2.5 beta 1)

### ISSUE THREAD

**Issue by awqk** [vkbo/novelWriter#1929]
1. open Manuscript Build Settings
2. change something on 1st page
3. close Manuscript Build Settings
4. question: save yes/no

1. open Manuscript Build Settings
2. go to any other page and change something
3. close Manuscript Build Settings
4. changes are silently discarded

It's also possible to get the settings in an inconsistent state, though I am unable to reproduce this. I had a situation where the scene headers disappeared from the preview unless the alt scene headers were hidden.

**Comment by vkbo** [comment]
The filters are applied as-is. This is not a bug, but I agree it is not intuitive. The "Apply" only actually applies to the settings themselves.

If you make setting changes, and don't Apply or Save, and then close the whole Manuscript tool, the setting panels are auto-cancelled, that's correct. Also not a bug, but maybe again not intuitive.

Finally, the reason some settings are disappearing is because you are probably also opening the same project with another version of the app. This converts your build setting to an older format when you open them. If you opened the same project with, say, 2.4 RC 1 and 2.3 back in that release cycle, you would lose the Alt Scene Header setting for instance.

**Comment by awqk** [comment]
> The filters are applied as-is. This is not a bug, but I agree it is not intuitive. The "Apply" only actually applies to the settings themselves.
> If you make setting changes, and don't Apply or Save, and then close the whole Manuscript tool, the setting panels are auto-cancelled, that's correct. Also not a bug, but maybe again not intuitive.

So the first page (Selection) is not a setting, and therefore it isn't auto-cancelled, whereas the other pages are settings?
I agree this is not intuitive :flushed:.

> Finally, the reason some settings are disappearing is because you are probably also opening the same project with another version of the app. This converts your build setting to an older format when you open them. If you opened the same project with, say, 2.4 RC 1 and 2.3 back in that release cycle, you would lose the Alt Scene Header setting for instance.

Didn't. Just installed 2.5b1 over 2.4.3 and never returned. Perhaps my description was misleading: it wasn't the settings disappearing, it was the scene headers in the preview text that didn't show up unless the alt scene headers were hidden, as if there was some confusion whether the scene headers in the documents were alt scene headers or not. There were several sessions with 2.5b1 before I noticed this. I tried a lot of hiding and showing with a unexpected results. Problem only vanished after close+open.

**Comment by vkbo** [comment]
> So the first page (Selection) is not a setting, and therefore it isn't auto-cancelled, whereas the other pages are settings? I agree this is not intuitive 😳.

No, it's not a setting per-say. It's meant to be a dynamic filter mechanism to be used live while you build so you don't have to hit "Apply". I'm not sure how to distinguish this feature in the dialog, but I do want to redesign the build settings dialog a little, which may help make this distinction as a side effect.

> Didn't. Just installed 2.5b1 over 2.4.3 and never returned. Perhaps my description was misleading: it wasn't the settings disappearing, it was the scene headers in the preview text that didn't show up unless the alt scene headers were hidden, as if there was some confusion whether the scene headers in the documents were alt scene headers or not. There were several sessions with 2.5b1 before I noticed this. I tried a lot of hiding and showing with a unexpected results. Problem only vanished after close+open.

Ah, ok. There could be a bug there. I would need a point by point list to reproduce, as it isn't clear to me how this is triggered.

I plan to release 2.5 RC 1 very soon, so it would be good to get in any fixes before that for a better release candidate before the final release.

**Comment by awqk** [comment]
> Ah, ok. There could be a bug there. I would need a point by point list to reproduce, as it isn't clear to me how this is triggered.

That's were I am confused myself, it could have involved closing the build tool without closing the build settings first. Or it was a pilot error, me not hitting Apply before updating the preview, repeatedly (as you said, it's not intuitivie), though I'm sure I would have tried the Save button at some point, it's hard to overlook.

**Comment by vkbo** [comment]
Using Apply before updating preview would be pretty standard behaviour I would expect. There is no way around that unless I make all settings "live".

I can make the filter section work on a copy of the build settings and make them also activate on Apply/Save. At least then it would be more consistent, but you would have to hit apply before each build it you're, say, building multiple manuscripts in a multi-novel project, one for each novel. That's precisely why the filter is applied directly.

**Comment by awqk** [comment]
Sounds good. In a multi-novel project, I wouldn't have thought of using different filter settings on the fly to create the manuscripts. Instead, creating a build setting for each individual novel would have been my first idea.

**Comment by vkbo** [comment]
Well, the feature was built to satisfy a feature request that specifically related to building for individual novels, or even chapters or single documents. So the filtering was treated as a separate feature. In the original build settings layout they were even on a panel on their own next to the settings, instead of on a tab. However, that used up too much screen space, so I redesigned it.

I'm a little reluctant to abandon this use case, as I also use this myself, so let's see how it looks after I've restructured the settings dialog. Of course, pressing Apply isn't a huge disadvantage, so it doesn't prevent this use case.

**Comment by vkbo** [comment]
As of 2.7, the Build Settings dialog works on a copy of the current build settings, so changes are no longer applied directly and not until Apply or OK is clicked. See #2351.

I believe can be closed as completed.
